### PR TITLE
Add `vers` support in index_packages endpoint

### PR DESCRIPTION
**Pull request by keshav-space**
- resolves https://github.com/nexB/purldb/issues/155

### CODE REVIEW

**Review by pombredanne — COMMENTED** [review]
Thanks! this is looking good, but I would prefer we find a way to avoid code copy from VulnerableCode. How could we share this code?
And if we go ahead and merge this, what's the plan to remove this duplication?

**Comment by tdruez** [comment]
This PR is a blocker for https://github.com/nexB/scancode.io/pull/836#issuecomment-1692964810
@keshav-space @pombredanne what's the status here?

**Comment by keshav-space** [comment]
> This PR is a blocker for [nexB/scancode.io#836 (comment)](https://github.com/nexB/scancode.io/pull/836#issuecomment-1692964810) @keshav-space @pombredanne what's the status here?

@tdruez waiting for review from @pombredanne and @JonoYang

File: packagedb/package_managers.py [hunk 1]
```diff
@@ -0,0 +1,635 @@
...
+    RubyVersionAPI,
+    GoproxyVersionAPI,
+    NpmVersionAPI,
+    HexVersionAPI,
+    LaunchpadVersionAPI,
+    CratesVersionAPI,
+    DebianVersionAPI,
+    ConanVersionAPI,
+}
+
```
Comment (JonoYang) [inline packagedb/package_managers.py:598]: Add a new line here

File: packagedb/package_managers.py [hunk 2]
```diff
@@ -0,0 +1,635 @@
...
+            release_date = self.get_latest_date(download_items)
+            yield PackageVersion(
+                value=version,
+                release_date=release_date,
+            )
+
+    def get_latest_date(self, downloads):
+        """
+        Return the latest date from a list of mapping of PyPI ``downloadss`` or  None.
```
Comment (JonoYang) [inline packagedb/package_managers.py:208]:
```suggestion
        Return the latest date from a list of mapping of PyPI ``downloads`` or  None.
```

File: packagedb/package_managers.py [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+          },
+          {
+            ....
+            "upload_time_iso_8601": "2010-12-23T05:20:23.509436Z",
+            "url": "https://files.pythonhosted.org/packages/8f/1f/561bddc20ca80fa5df025cc/Django-1.1.3.wheel",
+          },
+        ]
+        """
+        latest_date = None
+        for download in downloads:
```
Comment (JonoYang) [inline packagedb/package_managers.py:225]: I think its more Pythonic to sort the list of downloads by the upload time, then popping the latest download entry from the list.

```suggestion
        # Sort `downloads` from earliest "upload_time_iso_8601" to latest
        downloads = downloads.sort(key=lambda x: x["upload_time_iso_8601"])
        # Pop the latest download entry from `downloads`
        latest_download = downloads.pop()
        latest_date = latest_download.get("upload_time_iso_8601")
        if latest_date:
            latest_date = dateparser.parse(latest_date)
        return latest_date
```

File: packagedb/api.py [hunk 4]
```diff
@@ -39,6 +39,14 @@
...
 from packagedb.serializers import PackageSetAPISerializer
 from packagedb.serializers import PartySerializer
+from packagedb.package_managers import get_api_package_name
+from packagedb.package_managers import get_version_fetcher
+from packagedb.package_managers import VERSION_API_CLASSES_BY_PACKAGE_TYPE
+
+from univers import versions
+from univers.version_range import RANGE_CLASS_BY_SCHEMES
+from univers.version_range import InvalidVersionRange
+from univers.version_range import VersionRange
```
Comment (JonoYang) [inline packagedb/api.py:50]: Add a new line here

File: packagedb/package_managers.py [hunk 5]
```diff
@@ -0,0 +1,635 @@
+#
```
Comment (JonoYang) [inline packagedb/package_managers.py:1]: Bring the tests from https://github.com/nexB/vulnerablecode/blob/main/vulnerabilities/tests/test_package_managers.py over

**Review by JonoYang — CHANGES_REQUESTED** [review]
@keshav-space Thanks for the PR! I've left some comments for your consideration.

File: packagedb/package_managers.py (hunk 3 above)
Comment (keshav-space) [reply packagedb/package_managers.py:225]: @JonoYang I was trying to keep `package_managers` code untouched and refactor it while migrating to fetchcode.

Comment (JonoYang) [reply packagedb/package_managers.py:225]: @keshav-space Ah, I see.

File: packagedb/api.py [hunk 6]
```diff
@@ -373,12 +393,19 @@ def index_packages(self, request, *args, **kwargs):
...
+            - A list of package urls that were not placed on the queue.
+        - unsupported_vers_count
+            - The number of vers range that are not supported by the univers or package_manager.
+        - unsupported_vers
+            - A list of vers range that are not supported by the univers or package_manager.
         """
-        purls = request.data.getlist('package_urls')
-        unique_purls = set(purls)
+
+        purls = request.data.get('packages')
```
Comment (pombredanne) [inline packagedb/api.py:404]: Make it more robust with:
```suggestion
        purls = request.data.get('packages') or []
```

File: packagedb/api.py [hunk 7]
```diff
@@ -644,3 +673,87 @@ def _get_enhanced_package(package, packages):
...
+        if resolved:= resolve_verse(parsed_purl, vers):
+            unique_resolved_purls.update(resolved)
+        else:
+            unsupported_vers.add(vers)
+
+    return list(unique_resolved_purls), list(unsupported_purls), list(unsupported_vers)
+
+
+
+def resolve_verse(parsed_purl, vers):
```
Comment (pombredanne) [inline packagedb/api.py:705]:
```suggestion
def resolve_versions(parsed_purl, vers):
```

File: packagedb/api.py [hunk 8]
```diff
@@ -644,3 +673,87 @@ def _get_enhanced_package(package, packages):
...
+    if not package_name or not versionAPI:
+        return    
+
+    all_versions = versionAPI().fetch(package_name)
+    versionClass = VERSION_CLASS_BY_PACKAGE_TYPE.get(purl.type)
+
+    return [versionClass(package_version.value) for package_version in all_versions]
+
+
+VERSION_CLASS_BY_PACKAGE_TYPE = {
```
Comment (pombredanne) [inline packagedb/api.py:748]: Can we get this from univers instead?
 Either patch univers to expose a VERSION_CLASS_BY_PACKAGE_TYPE or provide a function there that returns a version class for a type. This should in all cases be computed and not hardcoded.

If you use the built-in mapping and use the version class attribute you will get what you need
https://github.com/nexB/univers/blob/205da7ecbf7b0f195662373ea710b2b84a877eb0/src/univers/version_range.py#L417 
```suggestion
VERSION_CLASS_BY_PACKAGE_TYPE = {pkg_type: range_class.version_class for pkg_type, range_class in RANGE_CLASS_BY_SCHEMES.items()}
```

```

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks... some nits for your consideration.

**Review by JonoYang — APPROVED** [review]
@keshav-space LGTM! 

@pombredanne Is there anything else you see that needs to be updated?

**Comment by JonoYang** [comment]
@keshav-space 

Merging, thanks for the contribution!

**Pull request merged**